docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+of partitions that may be active, and force-sleep the least recently used
+partitions when we hit the limit, but that doesn't help if all the partitions
+in the system really are active (we'd just be flapping them in and out of
+hibernation).
+
+That nondeterministic footprint means this only makes sense at some
+statistical scale where the ratio of idle partitions can be proactively
+monitored and the system grown/shrunk in response.  For example, we might
+have a system that is in principle able to support 1M partitions across
+ten nodes, but when many are idle it might shrink right down to a 3 node
+cluster.  We would re-expand the cluster (either by hand or via some
+external infrastructure automation) if that idle ratio changed.
+
+### Latency to wake
+
+There will be a short delay on the first I/O to a partition that wakes
+it from hibernation.  On a healthy system this should be in the range
+of a few hundred milliseconds.
+
+## Alternatives
+
+### Do nothing
+
+The practical capacity of a system will be how many _active_ partitions
+we can handle: this has the advantage of determinism, as there is no
+risk of a system 
+
+### Radically improved resource management elsewhere
+
+Separately to this RFC, we know that various subsystems' resource management
+can be improved, such as the way we keep unbounded in-memory indices in
+the storage layer, and hold open file descriptors for segments.
+
+In the raft layer, hibernation would be less important if we moved to a
+model where several partitions were grouped together in one raft group,
+thereby avoiding the "heartbeat per partition" scaling issue.  This would
+be a large architectural change.
+
+### Only hibernate on explicit command from user
+
+This is viable where an external control plane has its own tracking
+of the most recent I/O on topics.  However, it requires a more
+chatty external protocol for the external control plane to learn
+the access times of partitions to figure out which ones should be
+shut down.
+
+As long as the policy for hibernating is as simple as a timeout, it
+makes sense to build it in.  We could optionally also add an API for
+hibernation so that 
+
+### Hibernate per-partition rather than per-topic
+
+There is a reasonable expectation that applications will produce
+and consume to topics rather than partitions, and per-topic hibernation
+is much simpler to expose in user interfaces, as well as writing less
+to the controller log to control it.
+
+A situation where some partitions in a topic are idle but others
+are active is a fringe case, and this feature is aimed at providing
+statistically meaningful efficiency improvements at scale: it's fine
+if we miss the opportunity to hibernate any "weird" partitions like that.
+
+### Allow reads from hibernated partitions
+
+If a partition is hibernated, its data is still present on disk, and it
+is safe to serve reads from any replica, because no writes are going
+on.
```

Comment:
I don't think this is universally true, but probably true in some important cases. For example, we wouldn't want to allow reads from an isolated replica if the other replicas had been woken up formed a quorum. Assuming they are all hibernated, we'd also need to ensure we hibernated with a consistent view of data on all replicas.

Being able to support this IMO suggests a more partition-oriented implementation of hibernation that integrates into raft itself.